### PR TITLE
Implement CctpAdapter

### PR DESCRIPTION
**Pull request by mellowcroc**

### CODE REVIEW

File: apps/contracts/scripts/test-cctp-adapter-transfer-remote.ts [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+const MIN_AMOUNT = BigNumber.from(1000); // 0.001 USDC
+
+async function transferRemote(
+  signer: ethers.Wallet,
+  core: HyperlaneCore,
+  multiProvider: MultiProvider,
+  addresses: any
+) {
+  // Need at least `transferAmount` on the first CctpAdapter contract defined in `artifacts/cctp-adapter-addresses.json`
```
Comment (nambrot) [inline apps/contracts/scripts/test-cctp-adapter-transfer-remote.ts]: What does this mean?

File: apps/contracts/scripts/test-cctp-adapter-transfer-remote.ts [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    )
+  ) {
+    throw new Error(`Insufficient funds on ${firstChainId}`);
+  }
+
+  // Distribute USDC to all chains first and use the distributed amount for testing transfer
+  const distributionAmount = MIN_AMOUNT.div(length - 1);
+  const transferAmount = distributionAmount.div(length - 1);
+  for (let i = 0; i < length; i++) {
+    const chainId = Object.keys(addresses)[i];
```
Comment (nambrot) [inline apps/contracts/scripts/test-cctp-adapter-transfer-remote.ts]: Nit: I think you could do these loops with `for (const chainId of Object.keys(addresses)` which IMO reads better

File: apps/contracts/src/CctpAdapterDeployer.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+      config.cctpMessageTransmitter,
+      config.offchainUrls,
+    ];
+  }
+
+  router(contracts: HyperlaneContracts<CctpAdapterFactories>): CctpAdapter {
+    return contracts.router;
+  }
+
+  async enrollRemoteRouters(
```
Comment (nambrot) [inline apps/contracts/src/CctpAdapterDeployer.ts]: Is this override neccessary?

File: apps/frontend/src/pages/api/attestations.ts [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    console.log("sender:", sender);
+
+    const destinationDomain = extractDestinationDomainFromMessage(data);
+    console.log("destinationDomain:", destinationDomain);
+    let graphQLEndpoint = "";
+    let cctpEndpoint = "";
+    switch (destinationDomain) {
+      case DOMAIN_ETH:
+        cctpEndpoint = CCTP_API_ENDPOINT_MAINNET;
+        graphQLEndpoint =
```
Comment (nambrot) [inline apps/frontend/src/pages/api/attestations.ts]: Probably move this into dedicated config?

File: apps/frontend/src/pages/api/attestations.ts [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+const extractDestinationDomainFromMessage = (msg: string) => {
+  let destinationDomain = msg.substring(12, 20);
+  destinationDomain = destinationDomain.replace(/^0+/, "");
+  destinationDomain = "0x" + destinationDomain;
+  return parseInt(destinationDomain);
+};
+
+const extractNonceFromMessage = (msg: string) => {
+  console.log("msg:", msg);
+  const nonce = msg.substring(156 + 240, 156 + 240 + 8 * 2); // 156 is offset to Hyperlane message body, 240 is offset to nonce
```
Comment (nambrot) [inline apps/frontend/src/pages/api/attestations.ts]: Move this into a constant at the top

File: apps/frontend/src/pages/api/attestations.ts [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        }
+      }`,
+      variables: {
+        nonce: nonce,
+      },
+    },
+  };
+
+  let cctpMsg = "";
+  await axios
```
Comment (nambrot) [inline apps/frontend/src/pages/api/attestations.ts:136]: Just use standard fetch?

File: apps/contracts/scripts/test-cctp-adapter-transfer-remote.ts (hunk 1 above)
Comment (mellowcroc) [reply apps/contracts/scripts/test-cctp-adapter-transfer-remote.ts]: Need at least `transferAmount` on the first *network* defined in `artifacts/cctp-adapter-addresses.json`
e.g.
```
{
  "goerli": "0x001C83d5364C3360a27D0a72833C70203a864893",
  "fuji": "0xe7C82ea77091239eC52DD3f7640A36D132e0D981",
  "optimismgoerli": "0x8F2Ec21F10dDD8d581f8957aB1F1D484F4875977",
  "arbitrumgoerli": "0x528f3d2c128471990eb2DA54488b421790db6A4c"
}
```

File: apps/contracts/scripts/test-cctp-adapter-transfer-remote.ts (hunk 2 above)
Comment (mellowcroc) [reply apps/contracts/scripts/test-cctp-adapter-transfer-remote.ts]: Done

File: apps/contracts/src/CctpAdapterDeployer.ts (hunk 3 above)
Comment (mellowcroc) [reply apps/contracts/src/CctpAdapterDeployer.ts]: removed

File: apps/frontend/src/pages/api/attestations.ts (hunk 4 above)
Comment (mellowcroc) [reply apps/frontend/src/pages/api/attestations.ts]: done

File: apps/frontend/src/pages/api/attestations.ts (hunk 5 above)
Comment (mellowcroc) [reply apps/frontend/src/pages/api/attestations.ts]: done

File: apps/frontend/src/pages/api/attestations.ts (hunk 6 above)
Comment (mellowcroc) [reply apps/frontend/src/pages/api/attestations.ts:136]: failed to make graphql query work with fetch for some reason 😂

**Pull request merged**